File: network_analytics.py
"""Advanced network analytics utilities."""

from __future__ import annotations

from typing import Iterable, Mapping, Any, List


def find_suspicious_aps(records: Iterable[Mapping[str, Any]]) -> List[Mapping[str, Any]]:
    """Return Wi-Fi access points that may be suspicious.

    Simple heuristics flag open networks or duplicate BSSIDs broadcasting
    multiple SSIDs.
    """
    aps: List[Mapping[str, Any]] = []
    seen_bssid: dict[str, set[str]] = {}
    for rec in records:
        bssid = rec.get("bssid")
        ssid = rec.get("ssid") or ""
        enc = rec.get("encryption") or ""
        if enc.lower() == "open":
            aps.append(rec)
        if bssid:
            seen_bssid.setdefault(bssid, set()).add(ssid)
            if len(seen_bssid[bssid]) > 1:
                aps.append(rec)
    return aps

File: webui/src/networkAnalytics.js
export function cachedLookupVendor(_bssid) {
  return 'Vendor';
}

export function findSuspiciousAps(records) {
  const aps = [];
  const seenBssid = {};
  for (const rec of records) {
    const bssid = rec.bssid;
    const ssid = rec.ssid || '';
    const enc = (rec.encryption || '').toLowerCase();
    const channel = rec.channel;
    let suspicious = false;
    if (enc.includes('open') || enc.includes('wep')) {
      suspicious = true;
    }
    if (bssid) {
      seenBssid[bssid] = seenBssid[bssid] || new Set();
      seenBssid[bssid].add(ssid);
      if (seenBssid[bssid].size > 1) suspicious = true;
    }
    if (channel !== undefined && channel !== null && channel !== '') {
      try {
        const ch = parseInt(String(channel).split(' ')[0], 10);
        if (isNaN(ch) || ch < 1 || ch > 196) suspicious = true;
      } catch {
        suspicious = true;
      }
    }
    if (bssid && cachedLookupVendor(bssid) == null) {
      suspicious = true;
    }
    if (suspicious) aps.push(rec);
  }
  return aps;
}
